docs/environments/jigsaw.md

```diff
@@ -0,0 +1,55 @@
+# Jigsaw Environment
+
+<p align="center">
+        <img src="../env_anim/jigsaw.gif" width="500"/>
+</p>
+
+We provide here a Jax JIT-able implementation of a simple _jigsaw_ puzzle. The goal of the agent is to place
+all the jigsaw pieces in the correct locations on an empty 2D puzzle board. Each time an episode resets a
+new puzzle and set of piece is created. Pieces are randomly shuffled and rotated.
+
+## Observation
+The observation given to the agent gives a view of the current state of the puzzle as well as
+all pieces that can be placed.
+
+- `current_board`: jax array (float32) of shape `(num_rows, num_cols)` with values in the range
+    `[1, num_pieces]` (corresponding to the number of each piece). This board will have zeros
+    where no pieces have been placed and numbers corresponding to each piece where that particular
+    pieces has been paced.
+
+- `pieces`: jax array (float32) of shape `(num_pieces, 3, 3)` of all possible pieces in the
+    current puzzle. These pieces are shuffled and rotated. Pieces will always have shape `(3, 3)`.
+
+- `action_mask`: jax array (bool) of shape `(num_pieces, 4, num_rows-3, num_cols-3)`, representing
+    which actions are possible given the current state of the board. The first index indicates the
+    number of pieces in a given puzzle. The second index indicates the number of times a piece may be rotated.
+    The third and fourth indices indicate the row and column coordinate of where a piece may be placed respectively.
+    These values will always be `num_rows-3` and `num_cols-3` respectively to make it impossible for an agent to
+    place a piece outside the current board.
+
+
+## Action
+The action space is a `MultiDiscreteArray`, specifically a tuple of an index between 0 and `num_pieces`,
+an index between 0 and 4 (since there are 4 possible rotations), an index between 0 and `num_rows-3`
+(the possible row coordinates for placing a piece) and an index between 0 and `num_cols-3`
```

Comment:
```suggestion
an index between 0 and 4 (since there are 4 possible rotations), an index between 0 and `num_rows-2`
(the possible row coordinates for placing a piece) and an index between 0 and `num_cols-2`
```